src/cpp/src/storage/checkpointer.cpp

```diff
@@ -15,6 +16,54 @@ Checkpointer::Checkpointer(std::shared_ptr<Model> model, shared_ptr<GraphModelSt
     config_ = config;
 }
 
+void Checkpointer::create_checkpoint(string checkpoint_dir, CheckpointMeta checkpoint_meta, int epochs, int frequency) {
+    int prev_checkpoint = epochs - frequency;
+    std::string prev_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(prev_checkpoint) + "/";
+
+    string tmp_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(epochs) + "_tmp/";
+    std::filesystem::create_directory(tmp_checkpoint_dir);
+    
+    std::string new_embeddings_file = tmp_checkpoint_dir
+                                        + PathConstants::embeddings_file 
+                                        + PathConstants::file_ext;
+    std::string new_embeddings_state_file = tmp_checkpoint_dir
+                                            + PathConstants::embeddings_state_file
+                                            + PathConstants::file_ext;
+
+    if(std::filesystem::exists(prev_checkpoint_dir)) {
+        std::string embeddings_file = prev_checkpoint_dir 
+                                        + PathConstants::embeddings_file 
+                                        + PathConstants::file_ext;
+        std::string embeddings_state_file = prev_checkpoint_dir 
+                                                + PathConstants::embeddings_state_file
+                                                + PathConstants::file_ext;
+
+        if(std::filesystem::exists(embeddings_file)) {
+            std::filesystem::rename(embeddings_file, new_embeddings_file);
+            std::filesystem::rename(embeddings_state_file, new_embeddings_state_file);
+        }
+    } else {
+        std::string embeddings_file = checkpoint_dir 
+                                        + PathConstants::embeddings_file
+                                        + PathConstants::file_ext;
+        std::string embeddings_state_file = checkpoint_dir 
+                                                + PathConstants::embeddings_state_file
+                                                + PathConstants::file_ext;
+
+        if(std::filesystem::exists(embeddings_file)) {
+            std::filesystem::copy(embeddings_file, new_embeddings_file);
+            std::filesystem::copy(embeddings_state_file, new_embeddings_state_file);
+        }
+    }
+
+    this->save(tmp_checkpoint_dir, checkpoint_meta);
+
+    string final_checkpoint_dir = checkpoint_dir + "checkpoint_" + std::to_string(epochs) + "/";
+    std::filesystem::rename(tmp_checkpoint_dir, final_checkpoint_dir);
+    
+    std::filesystem::remove_all(prev_checkpoint_dir);
```

Comment:
Not sure it makes sense to delete the previous checkpoint always? A user might want to keep their previous checkpoints around. 

Maybe we should have an option for this.